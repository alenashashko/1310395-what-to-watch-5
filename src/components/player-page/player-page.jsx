import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {withRouter} from 'react-router';
import {connect} from 'react-redux';

import proptypes from '../../type';
import {formatMovieDuration, checkMovieIsLoaded} from '../../utils';
import withPlayerManager from '../../hocs/with-player-manager/with-player-manager';
import {MS_IN_SEC, ABSENT_PROGRESS_IN_PERSENT, FULL_PROGRESS_IN_PERCENT} from '../../const';
import {fetchMovieByID} from '../../store/api-actions';
import {generateWithFetchedData} from '../../hocs/with-fetched-data/with-fetched-data';
import {getMovie} from '../../store/selectors';

class PlayerPage extends PureComponent {
  constructor(props) {
    super(props);

    this._onExitClick = this._onExitClick.bind(this);
  }

  _onExitClick() {
    const {movie, destroy, history} = this.props;

    destroy();

    history.push(`/films/${movie.id}`);
  }

  render() {
    const {
      movie,
      duration,
      playbackTime,
      isPlaying,
      containerForwardRef,
      videoForwardRef,
      onCanPlayThrough,
      onTimeUpdate,
      onVideoEnded,
      onPauseButtonClick,
      onPlayButtonClick,
      onFullScreenButtonClick,
      onProgressBarClick
    } = this.props;

    const msLeft = (duration - playbackTime) * MS_IN_SEC;
    const progressInPercent = duration === ABSENT_PROGRESS_IN_PERSENT ?
      ABSENT_PROGRESS_IN_PERSENT : (playbackTime / duration) * FULL_PROGRESS_IN_PERCENT;

    return (
      <div className="player" ref={containerForwardRef}>
        <video
          ref={videoForwardRef}
          onCanPlayThrough={onCanPlayThrough}
          onTimeUpdate={onTimeUpdate}
          onEnded={onVideoEnded}
          className="player__video"
          src={movie.src}
          poster={movie.backgroundPicture}
          autoPlay
        />

        <button
          onClick={this._onExitClick}
          type="button"
          className="player__exit">
            Exix
        </button>

        <div className="player__controls">
          <div className="player__controls-row">
            <div className="player__time">
              <progress
                onClick={onProgressBarClick}
                className="player__progress"
                value={progressInPercent}
                max="100"
              />
              <div
                className="player__toggler"
                style={{left: `${progressInPercent}%`}}>
                Toggler
              </div>
            </div>
            <div className="player__time-value">{formatMovieDuration(msLeft)}</div>
          </div>

          <div className="player__controls-row">
            {isPlaying ? (
              <button onClick={onPauseButtonClick} type="button" className="player__play">
                <svg viewBox="0 0 14 21" width="14" height="21">
                  <use xlinkHref="#pause"></use>
                </svg>
                <span>Pause</span>
              </button>
            ) : (
              <button onClick={onPlayButtonClick} type="button" className="player__play">
                <svg viewBox="0 0 19 19" width="19" height="19">
                  <use xlinkHref="#play-s"></use>
                </svg>
                <span>Play</span>
              </button>
            )}

            <div className="player__name">{movie.title}</div>

            <button
              onClick={onFullScreenButtonClick}
              type="button"
              className="player__full-screen">
              <svg viewBox="0 0 27 27" width="27" height="27">
                <use xlinkHref="#full-screen"></use>
              </svg>
              <span>Full screen</span>
            </button>
          </div>
        </div>
      </div>
    );
  }
}

PlayerPage.propTypes = {
  id: proptypes.id,
  movie: proptypes.movie,
  history: proptypes.history,
  containerForwardRef: PropTypes.object.isRequired,
  videoForwardRef: PropTypes.object.isRequired,
  duration: PropTypes.number.isRequired,
  playbackTime: PropTypes.number.isRequired,
  isPlaying: PropTypes.bool.isRequired,
  onFullScreenButtonClick: PropTypes.func.isRequired,
  onPlayButtonClick: PropTypes.func.isRequired,
  onPauseButtonClick: PropTypes.func.isRequired,
  onTimeUpdate: PropTypes.func.isRequired,
  onProgressBarClick: PropTypes.func.isRequired,
  onCanPlayThrough: PropTypes.func.isRequired,
  onVideoEnded: PropTypes.func.isRequired,
  destroy: PropTypes.func.isRequired,
  fetchMovieByIDAction: PropTypes.func.isRequired
};

const mapStateToProps = (state) => ({
  movie: getMovie(state)
});

const mapDispatchToProps = (dispatch) => ({
  fetchMovieByIDAction(id) {
    dispatch(fetchMovieByID(id));
  }
});

export {PlayerPage};
export default connect(mapStateToProps, mapDispatchToProps)(withPlayerManager(withRouter(
    generateWithFetchedData(
        (state, props) => checkMovieIsLoaded(state, props),
        (props) => props.fetchMovieByIDAction(props.id)
    )(PlayerPage)
))
);
