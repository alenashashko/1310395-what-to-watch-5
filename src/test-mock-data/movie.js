export const movie = {
  id: `1`,
  title: `Orlando`,
  picture: `/img/orlando.jpg`,
  backgroundPicture: `/img/orlando.jpg`,
  backgroundColor: `#e1b0b2`,
  poster: `/img/orlando.jpg`,
  genre: `Crime`,
  year: 2015,
  ratingScore: 7.8,
  ratingCount: 1000,
  src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  previewSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  duration: 110,
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Cras aliquet varius magna, non porta ligula feugiat eget.
  Fusce tristique felis at fermentum pharetra.
  Aliquam id orci ut lectus varius viverra.
  Nullam nunc ex, convallis sed finibus eget, sollicitudin eget ante.
  Phasellus eros mauris, condimentum sed nibh vitae, sodales efficitur ipsum.
  Sed blandit, eros vel aliquam faucibus, purus ex euismod diam, eu luctus nunc ante ut dui.
  Sed sed nisi sed augue convallis suscipit in sed felis.
  Aliquam erat volutpat.`,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`,
    `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`],
  isFavorite: false
};

export const movies = [{
  id: `1`,
  title: `Orlando`,
  picture: `/img/orlando.jpg`,
  backgroundPicture: `/img/orlando.jpg`,
  backgroundColor: `#e1b0b2`,
  poster: `/img/orlando.jpg`,
  genre: `Crime`,
  year: 2015,
  ratingScore: 7.8,
  ratingCount: 1000,
  src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  previewSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  duration: 110,
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Cras aliquet varius magna, non porta ligula feugiat eget.
  Fusce tristique felis at fermentum pharetra.
  Aliquam id orci ut lectus varius viverra.
  Nullam nunc ex, convallis sed finibus eget, sollicitudin eget ante.
  Phasellus eros mauris, condimentum sed nibh vitae, sodales efficitur ipsum.
  Sed blandit, eros vel aliquam faucibus, purus ex euismod diam, eu luctus nunc ante ut dui.
  Sed sed nisi sed augue convallis suscipit in sed felis.
  Aliquam erat volutpat.`,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`,
    `Saoirse Ronan`, `Tony Revoloru`, `Tilda Swinton`, `Tom Wilkinson`],
  isFavorite: false
},
{
  id: `2`,
  title: `Revenant`,
  picture: `/img/revenant.jpg`,
  backgroundPicture: `/img/revenant.jpg`,
  backgroundColor: `#bdad8f`,
  poster: `/img/revenant.jpg`,
  genre: `Romance`,
  year: 2010,
  ratingScore: 5.9,
  ratingCount: 850,
  src: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  previewSrc: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  duration: 145,
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
  Cras aliquet varius magna, non porta ligula feugiat eget.
  Fusce tristique felis at fermentum pharetra.
  Aliquam id orci ut lectus varius viverra.
  Nullam nunc ex, convallis sed finibus eget, sollicitudin eget ante.
  Phasellus eros mauris, condimentum sed nibh vitae, sodales efficitur ipsum.`,
  director: `Robert Zemeckis`,
  starring: [`Will Smith`, `Michael J. Fox`, `Harrison Ford`,
    `Sandra Bullock`, `Betty White`, `Tom Hanks`, `Denzel Washington`],
  isFavorite: false
}
];

export const genres = [`Comedies`, `Crime`, `Documentary`, `Dramas`, `Horror`,
  `Kids & Family`, `Romance`, `Sci-Fi`, `Thrillers`];

export const genre = `All genres`;
