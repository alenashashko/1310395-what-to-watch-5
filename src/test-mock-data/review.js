import {adaptCommentToClient} from '../services/adapters';

export const apiReviews = [
  {
    id: 1,
    user: {
      id: 11,
      name: `Jack`
    },
    rating: 2.4,
    comment: `It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie.`,
    date: `2020-11-03T13:38:44.769Z`
  },
  {
    id: 2,
    user: {
      id: 12,
      name: `Isaac`
    },
    rating: 7.2,
    comment: `A movie that will take you to another world full of emotions.`,
    date: `2020-10-23T13:38:44.769Z`
  }
];

export const review = adaptCommentToClient(apiReviews[0]);

export const reviews = apiReviews.map(adaptCommentToClient);

