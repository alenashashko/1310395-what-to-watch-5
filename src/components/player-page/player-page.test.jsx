import React from 'react';
import renderer from 'react-test-renderer';
import browserHistory from '../../browser-history';

import {PlayerPage} from './player-page';
import {movie} from '../../test-mock-data/movie';

const noop = () => {};
const SEC_IN_MIN = 60;

describe(`Should PlayerPage render correctly`, () => {
  it(`With playing video status`, () => {
    const tree = renderer
      .create(
          <PlayerPage
            id={`1`}
            movie={movie}
            history={browserHistory}
            containerForwardRef={React.createRef()}
            videoForwardRef={React.createRef()}
            duration={movie.duration * SEC_IN_MIN}
            playbackTime={0.20}
            isPlaying={true}
            onFullScreenButtonClick={noop}
            onPlayButtonClick={noop}
            onPauseButtonClick={noop}
            onTimeUpdate={noop}
            onProgressBarClick={noop}
            onCanPlayThrough={noop}
            onVideoEnded={noop}
            destroy={noop}
            fetchMovieByIDAction={noop}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`With paused video status`, () => {
    const tree = renderer
      .create(
          <PlayerPage
            id={`1`}
            movie={movie}
            history={browserHistory}
            containerForwardRef={React.createRef()}
            videoForwardRef={React.createRef()}
            duration={movie.duration * SEC_IN_MIN}
            playbackTime={0.20}
            isPlaying={false}
            onFullScreenButtonClick={noop}
            onPlayButtonClick={noop}
            onPauseButtonClick={noop}
            onTimeUpdate={noop}
            onProgressBarClick={noop}
            onCanPlayThrough={noop}
            onVideoEnded={noop}
            destroy={noop}
            fetchMovieByIDAction={noop}
          />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
