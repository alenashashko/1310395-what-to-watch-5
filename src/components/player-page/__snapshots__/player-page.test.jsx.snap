// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should PlayerPage render correctly With paused video status 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    onCanPlayThrough={[Function]}
    onEnded={[Function]}
    onTimeUpdate={[Function]}
    poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg"
    src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
  />
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          onClick={[Function]}
          value={0.0036231884057971015}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0.0036231884057971015%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        1:31:59
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="19"
          viewBox="0 0 19 19"
          width="19"
        >
          <use
            xlinkHref="#play-s"
          />
        </svg>
        <span>
          Play
        </span>
      </button>
      <div
        className="player__name"
      >
        Bronson
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Should PlayerPage render correctly With playing video status 1`] = `
<div
  className="player"
>
  <video
    autoPlay={true}
    className="player__video"
    onCanPlayThrough={[Function]}
    onEnded={[Function]}
    onTimeUpdate={[Function]}
    poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg"
    src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
  />
  <button
    className="player__exit"
    onClick={[Function]}
    type="button"
  >
    Exit
  </button>
  <div
    className="player__controls"
  >
    <div
      className="player__controls-row"
    >
      <div
        className="player__time"
      >
        <progress
          className="player__progress"
          max="100"
          onClick={[Function]}
          value={0.0036231884057971015}
        />
        <div
          className="player__toggler"
          style={
            Object {
              "left": "0.0036231884057971015%",
            }
          }
        >
          Toggler
        </div>
      </div>
      <div
        className="player__time-value"
      >
        1:31:59
      </div>
    </div>
    <div
      className="player__controls-row"
    >
      <button
        className="player__play"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="21"
          viewBox="0 0 14 21"
          width="14"
        >
          <use
            xlinkHref="#pause"
          />
        </svg>
        <span>
          Pause
        </span>
      </button>
      <div
        className="player__name"
      >
        Bronson
      </div>
      <button
        className="player__full-screen"
        onClick={[Function]}
        type="button"
      >
        <svg
          height="27"
          viewBox="0 0 27 27"
          width="27"
        >
          <use
            xlinkHref="#full-screen"
          />
        </svg>
        <span>
          Full screen
        </span>
      </button>
    </div>
  </div>
</div>
`;
